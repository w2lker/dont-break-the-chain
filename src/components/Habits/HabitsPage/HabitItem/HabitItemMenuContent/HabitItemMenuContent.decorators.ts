import withStyles from '@material-ui/core/styles/withStyles';
import HabitItemMenuContent from './HabitItemMenuContent';
import habitItemMenuContentStyles from './HabitItemMenuContent.styles';

const styledComponent = withStyles(habitItemMenuContentStyles)(HabitItemMenuContent);

export default styledComponent;
