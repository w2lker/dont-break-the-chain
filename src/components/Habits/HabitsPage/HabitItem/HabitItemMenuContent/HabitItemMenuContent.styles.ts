import { Styles } from '@material-ui/styles';
import { bon } from '../../../../../utils/styleUtils';

const habitItemMenuContentStyles: Styles<any, any> = {
  root: {
    ...bon.setFlex('space-between', 'center'),
  },
  icon: {
    ...bon.size(40),
    textAlign: 'center',
  },
  text: {
    paddingLeft: 15,
    width: 165,
  },
};

export default habitItemMenuContentStyles;
