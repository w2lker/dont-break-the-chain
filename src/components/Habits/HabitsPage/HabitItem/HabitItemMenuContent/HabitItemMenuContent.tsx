import React from 'react';

import { MenuItem } from '@material-ui/core';
import { WithStyles } from 'react-jss';
import { Link } from 'react-router-dom';

import MaterialDesignIcon from '../../../../basic/Material-Icon';
import habitItemMenuContentStyles from './HabitItemMenuContent.styles';

export interface IHabitItemMenuContentProps extends WithStyles<typeof habitItemMenuContentStyles> {
  icon: string;
  title: string;
  action?: () => void;
  url?: string;
}

class HabitItemMenuContent extends React.Component<IHabitItemMenuContentProps, {}> {
  render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
    const { icon, classes, title, action, url } = this.props;
    const menuContent = (
      <div className={classes.root}>
        <MaterialDesignIcon
          className={classes.icon}
          name={icon}
        />
        <span className={classes.text}>
        {title}
      </span>
      </div>
    );

    if (action) {
      return (
        <MenuItem onClick={action}>
          {menuContent}
        </MenuItem>
      );
    }

    return (
      <MenuItem
        component={Link as React.ElementType}
        to={url}
      >
        {menuContent}
      </MenuItem>
    );
  }
}

export default HabitItemMenuContent;
