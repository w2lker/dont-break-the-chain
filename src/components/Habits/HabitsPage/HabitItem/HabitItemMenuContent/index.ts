import HabitItemMenuContent from './HabitItemMenuContent.decorators';

export default HabitItemMenuContent;
