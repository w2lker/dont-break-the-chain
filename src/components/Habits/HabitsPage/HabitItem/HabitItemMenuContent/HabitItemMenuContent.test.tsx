import React from 'react';

import { mount, shallow } from 'enzyme';

import { MemoryRouter } from 'react-router-dom';
import HabitItemMenuContent, { IHabitItemMenuContentProps } from './HabitItemMenuContent';
import HabitItemMenuContentDecorated from './HabitItemMenuContent.decorators';
import habitItemMenuContentStyles from './HabitItemMenuContent.styles';

const setupTests = () => {
  const classes = {
    root: 'habitItemMenuContent-wrapper',
    icon: 'habitItemMenuContent-icon',
    text: 'habitItemMenuContent-text',
  };

  const classesKeys = Object.keys(classes);

  const sampleProps: IHabitItemMenuContentProps = {
    classes,
    icon: 'sample-icon',
    title: 'sample-text',
    action: jest.fn(),
  };

  const sampleLinkProps: IHabitItemMenuContentProps = {
    ...sampleProps,
    action: undefined,
    url: 'some-link',
  };

  const wrappedComponent = (props: IHabitItemMenuContentProps) => (
    <MemoryRouter>
      <HabitItemMenuContent {...props}/>
    </MemoryRouter>
  );

  return { classes, classesKeys, sampleProps, sampleLinkProps, wrappedComponent };
};

describe('HabitItemMenuContent component', () => {

  it('match snapshot', () => {
    const { sampleProps, sampleLinkProps, wrappedComponent } = setupTests();
    const componentWithAction = mount(wrappedComponent(sampleProps));
    expect(componentWithAction.debug()).toMatchSnapshot();
    const componentWithLink = mount(wrappedComponent(sampleLinkProps));
    expect(componentWithLink.debug()).toMatchSnapshot();
  });

  it('renders icon', () => {
    const { sampleProps, classes } = setupTests();
    const component = shallow(<HabitItemMenuContent {...sampleProps} />);
    const iconProps = component.find('MaterialDesignIcon').props();
    expect(iconProps.className).toBe(classes.icon);
  });

  it('renders title', () => {
    const { sampleProps, classes } = setupTests();
    const component = shallow(<HabitItemMenuContent {...sampleProps} />);
    const title = component.find(`.${classes.text}`);
    expect(title.length).toBe(1);
    const content = title.props().children;
    expect(content).toBe(sampleProps.title);
  });

  it('perform click', () => {
    const { sampleProps } = setupTests();
    const component = shallow(<HabitItemMenuContent {...sampleProps} />);
    const menuItem = component.find('WithStyles(ForwardRef(MenuItem))');
    expect(menuItem.length).toBe(1);
    menuItem.simulate('click');
    expect(sampleProps.action).toBeCalledTimes(1);
  });

  it('creates link', () => {
    const { sampleLinkProps, wrappedComponent } = setupTests();
    const component = mount(wrappedComponent(sampleLinkProps));
    const link = component.find('Link');
    const dest = link.props().to;
    expect(dest).toBe(sampleLinkProps.url);
  });
});

describe('HabitItemMenuContent styles', () => {
  it('contains principle classes', () => {
    const { classesKeys } = setupTests();
    expect(habitItemMenuContentStyles).toBeDefined();
    classesKeys.forEach((keyValue) => {
      // @ts-ignore
      expect(habitItemMenuContentStyles[keyValue]).toBeDefined();
    });
  });
});

describe('HabitItemMenuContent decorators', () => {

  it('provides styled classes from decorators', () => {
    const { sampleProps, classesKeys } = setupTests();
    const props = {
      ...sampleProps,
      classes: undefined,
    };

    const component = mount(<HabitItemMenuContentDecorated {...props} />);
    // @ts-ignore
    const assignedClasses = component.find('HabitItemMenuContent').props().classes;
    classesKeys.forEach(keyValue => {
      expect(assignedClasses[keyValue]).toBeDefined();
    });
  });
});
