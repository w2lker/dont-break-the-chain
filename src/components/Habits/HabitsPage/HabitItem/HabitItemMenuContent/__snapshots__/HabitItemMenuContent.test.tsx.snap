// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitItemMenuContent component match snapshot 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <HabitItemMenuContent classes={{...}} icon=\\"sample-icon\\" title=\\"sample-text\\" action={[Function: mockConstructor]}>
      <WithStyles(ForwardRef(MenuItem)) onClick={[Function: mockConstructor]}>
        <ForwardRef(MenuItem) classes={{...}} onClick={[Function: mockConstructor]}>
          <WithStyles(ForwardRef(ListItem)) button={true} role=\\"menuitem\\" tabIndex={-1} component=\\"li\\" selected={[undefined]} disableGutters={false} classes={{...}} className=\\"MuiMenuItem-root MuiMenuItem-gutters\\" onClick={[Function: mockConstructor]}>
            <ForwardRef(ListItem) classes={{...}} button={true} role=\\"menuitem\\" tabIndex={-1} component=\\"li\\" selected={[undefined]} disableGutters={false} className=\\"MuiMenuItem-root MuiMenuItem-gutters\\" onClick={[Function: mockConstructor]}>
              <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" disabled={false} role=\\"menuitem\\" tabIndex={-1} onClick={[Function: mockConstructor]} component=\\"li\\" focusVisibleClassName=\\"Mui-focusVisible\\">
                <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" disabled={false} role=\\"menuitem\\" tabIndex={-1} onClick={[Function: mockConstructor]} component=\\"li\\" focusVisibleClassName=\\"Mui-focusVisible\\">
                  <li className=\\"MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" onBlur={[Function]} onClick={[Function: mockConstructor]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={-1} role=\\"menuitem\\" aria-disabled={false}>
                    <div className=\\"habitItemMenuContent-wrapper\\">
                      <MaterialDesignIcon className=\\"habitItemMenuContent-icon\\" name=\\"sample-icon\\">
                        <span className=\\"material-icon mdi mdi-sample-icon habitItemMenuContent-icon\\" style={[undefined]} />
                      </MaterialDesignIcon>
                      <span className=\\"habitItemMenuContent-text\\">
                        sample-text
                      </span>
                    </div>
                    <NoSsr>
                      <WithStyles(undefined) center={false}>
                        <ForwardRef(TouchRipple) classes={{...}} center={false}>
                          <span className=\\"MuiTouchRipple-root\\">
                            <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                          </span>
                        </ForwardRef(TouchRipple)>
                      </WithStyles(undefined)>
                    </NoSsr>
                  </li>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(ListItem)>
          </WithStyles(ForwardRef(ListItem))>
        </ForwardRef(MenuItem)>
      </WithStyles(ForwardRef(MenuItem))>
    </HabitItemMenuContent>
  </Router>
</MemoryRouter>"
`;

exports[`HabitItemMenuContent component match snapshot 2`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <HabitItemMenuContent classes={{...}} icon=\\"sample-icon\\" title=\\"sample-text\\" action={[undefined]} url=\\"some-link\\">
      <WithStyles(ForwardRef(MenuItem)) component={{...}} to=\\"some-link\\">
        <ForwardRef(MenuItem) classes={{...}} component={{...}} to=\\"some-link\\">
          <WithStyles(ForwardRef(ListItem)) button={true} role=\\"menuitem\\" tabIndex={-1} component={{...}} selected={[undefined]} disableGutters={false} classes={{...}} className=\\"MuiMenuItem-root MuiMenuItem-gutters\\" to=\\"some-link\\">
            <ForwardRef(ListItem) classes={{...}} button={true} role=\\"menuitem\\" tabIndex={-1} component={{...}} selected={[undefined]} disableGutters={false} className=\\"MuiMenuItem-root MuiMenuItem-gutters\\" to=\\"some-link\\">
              <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" disabled={false} role=\\"menuitem\\" tabIndex={-1} to=\\"some-link\\" component={{...}} focusVisibleClassName=\\"Mui-focusVisible\\">
                <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" disabled={false} role=\\"menuitem\\" tabIndex={-1} to=\\"some-link\\" component={{...}} focusVisibleClassName=\\"Mui-focusVisible\\">
                  <Link className=\\"MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" onBlur={[Function]} onClick={[undefined]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={-1} role=\\"menuitem\\" aria-disabled={false} to=\\"some-link\\">
                    <LinkAnchor className=\\"MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" onBlur={[Function]} onClick={[undefined]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={-1} role=\\"menuitem\\" aria-disabled={false} href=\\"/some-link\\" navigate={[Function: navigate]}>
                      <a className=\\"MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button\\" onBlur={[Function]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={-1} role=\\"menuitem\\" aria-disabled={false} href=\\"/some-link\\" onClick={[Function: onClick]}>
                        <div className=\\"habitItemMenuContent-wrapper\\">
                          <MaterialDesignIcon className=\\"habitItemMenuContent-icon\\" name=\\"sample-icon\\">
                            <span className=\\"material-icon mdi mdi-sample-icon habitItemMenuContent-icon\\" style={[undefined]} />
                          </MaterialDesignIcon>
                          <span className=\\"habitItemMenuContent-text\\">
                            sample-text
                          </span>
                        </div>
                        <NoSsr>
                          <WithStyles(undefined) center={false}>
                            <ForwardRef(TouchRipple) classes={{...}} center={false}>
                              <span className=\\"MuiTouchRipple-root\\">
                                <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                              </span>
                            </ForwardRef(TouchRipple)>
                          </WithStyles(undefined)>
                        </NoSsr>
                      </a>
                    </LinkAnchor>
                  </Link>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(ListItem)>
          </WithStyles(ForwardRef(ListItem))>
        </ForwardRef(MenuItem)>
      </WithStyles(ForwardRef(MenuItem))>
    </HabitItemMenuContent>
  </Router>
</MemoryRouter>"
`;
